File: helper/ohlcv_url_chart.html/ohlcv_url.py
#!/usr/bin/env python3
"""
ohlcv_url.py - Generate a shareable OHLCV-chart URL from trading data.

Produces the exact #d=... payload understood by ``ohlcv_url_chart.html`` so the
whole dataset travels inside the link (no server). The data pipeline mirrors the
web app:

    1. Columnar binary packing
         - regular timestamps  -> start + interval + count
           irregular           -> start + zigzag-varint deltas (epoch seconds)
         - O/H/L/C quantized to integer ticks (x10^decimals, lossless),
           then per-column delta -> zigzag -> varint
         - volume quantized + varint (raw or delta, whichever is smaller)
    2. raw DEFLATE  (zlib, wbits=-15)  == browser CompressionStream('deflate-raw')
    3. URL-safe text in the hash fragment:
         - default : Base64url
         - dense   : custom 73-char URL-fragment-safe alphabet (~3-5% shorter)

Use as a module:

    from ohlcv_url import generate_url_from_file, generate_url
    url = generate_url_from_file("data.csv")                 # default root URL
    url = generate_url_from_file("data.csv", root_url="https://host/x.html",
                                 dense=True, symbol="AAPL")
    url = generate_url(timestamps, opens, highs, lows, closes, volumes)

Use as a CLI:

    python ohlcv_url.py data.csv
    python ohlcv_url.py data.csv --root https://host/ohlcv_url_chart.html --dense
    python ohlcv_url.py data.csv --symbol AAPL --open      # also open in browser

Accepted CSV headers (case-insensitive):
    time   : timestamp | datetime | dts | date | time
    prices : open|o, high|h, low|l, close|c
    volume : volume | vol | v          (optional)
    timestamps may be ISO dates, "YYYY-MM-DD HH:MM:SS", or epoch seconds/ms.
    Naive (no-timezone) timestamps are interpreted as UTC.
"""

from __future__ import annotations

import math
import zlib
from datetime import datetime, timezone
from typing import Iterable, List, Optional, Sequence

DEFAULT_ROOT_URL = "https://abelho.github.io/util_many/ohlcv_url_chart.html"

# 73-char alphabet: all RFC-3986 fragment-safe and free of the '&'/'=' hash
# separators. Must match DENSE_ALPHABET in ohlcv_url_chart.html exactly.
DENSE_ALPHABET = ("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
                  "abcdefghijklmnopqrstuvwxyz-._~!$*+/:@")
_DENSE_N = len(DENSE_ALPHABET)

_TIME_ALIASES = ("timestamp", "datetime", "dts", "date", "time")
_OPEN_ALIASES = ("open", "o")
_HIGH_ALIASES = ("high", "h")
_LOW_ALIASES = ("low", "l")
_CLOSE_ALIASES = ("close", "c")
_VOL_ALIASES = ("volume", "vol", "v")


# --------------------------------------------------------------------------- #
# Low-level integer codecs (match the JS implementation)
# --------------------------------------------------------------------------- #
def _write_varint(out: bytearray, n: int) -> None:
    """Unsigned LEB128."""
    n = int(n)
    while n >= 128:
        out.append((n % 128) + 128)
        n //= 128
    out.append(n)


def _zigzag(n: int) -> int:
    return n * 2 if n >= 0 else n * -2 - 1


def _round_half_up(x: float) -> int:
    """Match JS Math.round (round half toward +Infinity)."""
    return math.floor(x + 0.5)


# --------------------------------------------------------------------------- #
# Stage 1 - columnar binary packing
# --------------------------------------------------------------------------- #
def _pack_binary(ts: Sequence[int], o, h, l, c, vol,
                 price_decimals: int, vol_decimals: int,
                 has_volume: bool, vol_float: bool) -> bytes:
    n = len(ts)
    pscale = 10 ** price_decimals
    vscale = 10 ** vol_decimals

    ts_regular = True
    interval = ts[1] - ts[0] if n >= 2 else 0
    for i in range(2, n):
        if ts[i] - ts[i - 1] != interval:
            ts_regular = False
            break

    out = bytearray()
    out += b"O1"          # magic
    out.append(1)         # version
    flags_idx = len(out)
    out.append(0)         # flags placeholder
    out.append(price_decimals & 0xFF)
    out.append(vol_decimals & 0xFF)
    _write_varint(out, n)

    if ts_regular:
        _write_varint(out, ts[0] if n else 0)
        _write_varint(out, interval if interval > 0 else 0)
    else:
        _write_varint(out, ts[0])
        for i in range(1, n):
            _write_varint(out, _zigzag(ts[i] - ts[i - 1]))

    for col in (o, h, l, c):
        prev = 0
        for i in range(n):
            v = _round_half_up(col[i] * pscale)
            _write_varint(out, _zigzag(v - prev))
            prev = v

    vol_delta = False
    if has_volume:
        scaled = [_round_half_up((vol[i] or 0) * vscale) for i in range(n)]
        raw = bytearray()
        delta = bytearray()
        prev = 0
        for i in range(n):
            _write_varint(raw, scaled[i])
            _write_varint(delta, _zigzag(scaled[i] - prev))
            prev = scaled[i]
        if len(delta) < len(raw):
            vol_delta = True
            out += delta
        else:
            out += raw

    flags = 0
    if ts_regular:
        flags |= 1
    if has_volume:
        flags |= 2
    if vol_float:
        flags |= 4
    if vol_delta:
        flags |= 8
    out[flags_idx] = flags
    return bytes(out)


# --------------------------------------------------------------------------- #
# Stage 2 - raw DEFLATE
# --------------------------------------------------------------------------- #
def _deflate_raw(data: bytes) -> bytes:
    co = zlib.compressobj(9, zlib.DEFLATED, -15)
    return co.compress(data) + co.flush()


# --------------------------------------------------------------------------- #
# Stage 3 - URL-safe text encodings
# --------------------------------------------------------------------------- #
def _b64url_encode(data: bytes) -> str:
    import base64
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _dense_encode(data: bytes) -> str:
    lead = 0
    while lead < len(data) and data[lead] == 0:
        lead += 1
    num = int.from_bytes(data[lead:], "big") if lead < len(data) else 0
    digits = ""
    while num > 0:
        digits = DENSE_ALPHABET[num % _DENSE_N] + digits
        num //= _DENSE_N
    return DENSE_ALPHABET[lead // _DENSE_N] + DENSE_ALPHABET[lead % _DENSE_N] + digits


# --------------------------------------------------------------------------- #
# Parsing / detection helpers
# --------------------------------------------------------------------------- #
def _decimals_of(token: str) -> int:
    token = token.strip()
    if "." not in token:
        return 0
    frac = token.rsplit(".", 1)[1]
    return len(frac) if frac.isdigit() else 0


def _detect_decimals_floats(values: Iterable[float]) -> int:
    d = 0
    for v in values:
        k = 0
        while k < 9 and abs(v * (10 ** k) - round(v * (10 ** k))) > 1e-9:
            k += 1
        d = max(d, k)
    return d


def to_epoch_seconds(value) -> int:
    """Accept int/float (s or ms), datetime, or string; return epoch seconds (UTC)."""
    if isinstance(value, datetime):
        dt = value if value.tzinfo else value.replace(tzinfo=timezone.utc)
        return int(round(dt.timestamp()))
    if isinstance(value, (int, float)):
        n = float(value)
        return int(round(n / 1000.0)) if abs(n) >= 1e12 else int(round(n))
    s = str(value).strip()
    if s.isdigit():
        return int(s) // 1000 if len(s) >= 12 else int(s)
    t = s.replace(" ", "T", 1)
    utc = False
    if t[-1:] in ("Z", "z"):
        t = t[:-1]
        utc = True
    try:
        dt = datetime.fromisoformat(t)
    except ValueError:
        dt = datetime.fromisoformat(t.split("T")[0])
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    elif utc:
        dt = dt.replace(tzinfo=timezone.utc)
    return int(round(dt.timestamp()))


def _find_col(header: List[str], aliases: Sequence[str]) -> int:
    for i, h in enumerate(header):
        if h in aliases:
            return i
    return -1


# --------------------------------------------------------------------------- #
# Public API
# --------------------------------------------------------------------------- #
def _encode(ts, o, h, l, c, vol, price_decimals, vol_decimals,
            has_volume, vol_float, root_url, dense, symbol) -> str:
    from urllib.parse import quote
    binary = _pack_binary(ts, o, h, l, c, vol, price_decimals, vol_decimals,
                          has_volume, vol_float)
    comp = _deflate_raw(binary)
    payload = _dense_encode(comp) if dense else _b64url_encode(comp)
    frag = "#d=" + payload
    if dense:
        frag += "&e=1"
    if symbol:
        frag += "&s=" + quote(str(symbol), safe="")
    return (root_url or DEFAULT_ROOT_URL) + frag


def generate_url(timestamps: Sequence,
                 opens: Sequence[float], highs: Sequence[float],
                 lows: Sequence[float], closes: Sequence[float],
                 volumes: Optional[Sequence[float]] = None,
                 root_url: str = DEFAULT_ROOT_URL,
                 dense: bool = False,
                 symbol: Optional[str] = None,
                 price_decimals: Optional[int] = None,
                 vol_decimals: Optional[int] = None) -> str:
    """Build a chart URL from in-memory OHLCV arrays.

    timestamps may be epoch seconds/ms, datetime objects, or strings.
    Decimal precision is auto-detected from the values unless given.
    """
    ts = [to_epoch_seconds(t) for t in timestamps]
    n = len(ts)
    if not (len(opens) == len(highs) == len(lows) == len(closes) == n):
        raise ValueError("All OHLC arrays must have the same length as timestamps.")
    has_volume = volumes is not None
    if has_volume and len(volumes) != n:
        raise ValueError("volumes must match the number of rows.")

    if price_decimals is None:
        price_decimals = _detect_decimals_floats(
            list(opens) + list(highs) + list(lows) + list(closes))
    price_decimals = max(0, min(price_decimals, 9))

    vol_float = False
    if vol_decimals is None and has_volume:
        vol_decimals = _detect_decimals_floats(volumes)
    vol_decimals = max(0, min(vol_decimals or 0, 9))
    if has_volume:
        vol_float = vol_decimals > 0

    return _encode(ts, opens, highs, lows, closes, volumes,
                   price_decimals, vol_decimals, has_volume, vol_float,
                   root_url, dense, symbol)


def parse_csv_file(path: str):
    """Return (ts, o, h, l, c, vol|None, price_decimals, vol_decimals,
    has_volume, vol_float) detected from a CSV file (string-accurate decimals)."""
    import csv
    with open(path, newline="", encoding="utf-8-sig") as fh:
        rows = [r for r in csv.reader(fh) if any(cell.strip() for cell in r)]
    if len(rows) < 2:
        raise ValueError("CSV has no data rows.")
    header = [c.strip().lower() for c in rows[0]]
    ti = _find_col(header, _TIME_ALIASES)
    oi = _find_col(header, _OPEN_ALIASES)
    hi = _find_col(header, _HIGH_ALIASES)
    li = _find_col(header, _LOW_ALIASES)
    ci = _find_col(header, _CLOSE_ALIASES)
    if ti < 0:
        raise ValueError("No time column (timestamp/datetime/dts/date/time).")
    if min(oi, hi, li, ci) < 0:
        raise ValueError("Need open, high, low and close columns.")
    vi = _find_col(header, _VOL_ALIASES)
    has_volume = vi >= 0

    need = max(ti, oi, hi, li, ci, vi if has_volume else 0)
    data = [r for r in rows[1:] if len(r) > need and r[ti].strip() != ""]
    if not data:
        raise ValueError("CSV has no usable data rows.")

    price_decimals = 0
    for r in data:
        for idx in (oi, hi, li, ci):
            price_decimals = max(price_decimals, _decimals_of(r[idx]))
    price_decimals = min(price_decimals, 9)

    vol_decimals = 0
    vol_float = False
    if has_volume:
        for r in data:
            d = _decimals_of(r[vi])
            if d > 0:
                vol_float = True
            vol_decimals = max(vol_decimals, d)
        vol_decimals = min(vol_decimals, 9)

    ts, o, h, l, c = [], [], [], [], []
    vol = [] if has_volume else None
    for r in data:
        ts.append(to_epoch_seconds(r[ti]))
        o.append(float(r[oi]))
        h.append(float(r[hi]))
        l.append(float(r[li]))
        c.append(float(r[ci]))
        if has_volume:
            try:
                vol.append(float(r[vi]))
            except ValueError:
                vol.append(0.0)
    return ts, o, h, l, c, vol, price_decimals, vol_decimals, has_volume, vol_float


def generate_url_from_file(path: str,
                           root_url: str = DEFAULT_ROOT_URL,
                           dense: bool = False,
                           symbol: Optional[str] = None) -> str:
    """Build a chart URL from a CSV file. Decimals are detected from the text
    (so the URL is byte-faithful to the file's precision)."""
    (ts, o, h, l, c, vol, price_decimals, vol_decimals,
     has_volume, vol_float) = parse_csv_file(path)
    if symbol is None:
        import os
        symbol = os.path.splitext(os.path.basename(path))[0]
    return _encode(ts, o, h, l, c, vol, price_decimals, vol_decimals,
                   has_volume, vol_float, root_url, dense, symbol)


# --------------------------------------------------------------------------- #
# CLI
# --------------------------------------------------------------------------- #
def _main(argv: Optional[List[str]] = None) -> int:
    import argparse
    parser = argparse.ArgumentParser(
        description="Generate a shareable OHLCV-chart URL from a CSV file.")
    parser.add_argument("csv", help="Path to an OHLCV CSV file.")
    parser.add_argument("--root", default=DEFAULT_ROOT_URL,
                        help="Root chart URL (default: %(default)s)")
    parser.add_argument("--dense", action="store_true",
                        help="Use the max-density URL encoding (~3-5%% shorter).")
    parser.add_argument("--symbol", default=None,
                        help="Chart title/symbol (default: CSV filename).")
    parser.add_argument("--open", action="store_true",
                        help="Open the generated URL in the default browser.")
    args = parser.parse_args(argv)

    url = generate_url_from_file(args.csv, root_url=args.root,
                                 dense=args.dense, symbol=args.symbol)
    print(url)
    if args.open:
        import webbrowser
        webbrowser.open(url)
    return 0


if __name__ == "__main__":
    raise SystemExit(_main())

File: helper/ohlcv_url_chart.html/OhlcvUrl.cs
// ===========================================================================
// OhlcvUrl.cs - Generate a shareable OHLCV-chart URL from trading data.
//
// Produces the exact #d=... payload understood by ohlcv_url_chart.html, so the
// whole dataset travels inside the link (no server). Pipeline mirrors the web
// app and the Python helper (helper/ohlcv_url.py):
//   1. Columnar binary packing (regular timestamps -> start+interval+count;
//      O/H/L/C quantized to integer ticks then delta -> zigzag -> varint;
//      volume varint, raw or delta).
//   2. raw DEFLATE  (System.IO.Compression.DeflateStream == browser
//      CompressionStream('deflate-raw')).
//   3. URL-safe text: Base64url (default) or a 73-char max-density alphabet.
//
// DEPENDENCIES: only the .NET BCL (System, System.IO.Compression). No
// System.Numerics needed - the max-density encoder uses manual big-number
// division so it compiles cleanly inside NinjaTrader 8 (.NET 4.8) NinjaScript.
//
// ------------------------------ USAGE --------------------------------------
// As a library (e.g. from a NinjaScript indicator/strategy):
//
//     using OhlcvUrlShare;
//     // Build arrays from the chart's bars (oldest -> newest):
//     int n = CurrentBar + 1;
//     var t = new DateTime[n]; var o = new double[n]; var h = new double[n];
//     var l = new double[n];   var c = new double[n]; var v = new double[n];
//     for (int i = 0; i < n; i++) {
//         int b = CurrentBar - i;                 // Bars are newest-first
//         t[n-1-i] = Time[b];  o[n-1-i] = Open[b]; h[n-1-i] = High[b];
//         l[n-1-i] = Low[b];   c[n-1-i] = Close[b]; v[n-1-i] = Volume[b];
//     }
//     string url = OhlcvUrl.GenerateUrl(t, o, h, l, c, v, symbol: Instrument.FullName);
//     Print(url);
//
// As a CLI (compile a console build with the OHLCV_CLI symbol defined):
//     csc /define:OHLCV_CLI OhlcvUrl.cs        ->  OhlcvUrl.exe data.csv [--root URL] [--dense] [--symbol NAME]
//
// Accepted CSV headers (case-insensitive):
//     time   : timestamp | datetime | dts | date | time
//     prices : open|o, high|h, low|l, close|c
//     volume : volume | vol | v          (optional)
//   Timestamps may be ISO dates, "YYYY-MM-DD HH:MM:SS", or epoch seconds/ms.
//   Naive (no-timezone) timestamps are interpreted as UTC.
// ===========================================================================

using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.IO.Compression;
using System.Text;

namespace OhlcvUrlShare
{
    /// <summary>Static utility that turns OHLCV data into a shareable chart URL.</summary>
    public static class OhlcvUrl
    {
        public const string DefaultRootUrl =
            "https://abelho.github.io/util_many/ohlcv_url_chart.html";

        // 73-char alphabet: RFC-3986 fragment-safe and free of the '&'/'='
        // hash separators. Must match DENSE_ALPHABET in ohlcv_url_chart.html.
        private const string DenseAlphabet =
            "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-._~!$*+/:@";
        private const int DenseN = 73; // == DenseAlphabet.Length

        // ------------------------------------------------------------------ //
        // Public API
        // ------------------------------------------------------------------ //

        /// <summary>Build a chart URL from in-memory arrays (timestamps as epoch seconds).</summary>
        public static string GenerateUrl(
            long[] epochSeconds, double[] open, double[] high, double[] low, double[] close,
            double[] volume = null, string rootUrl = null, bool dense = false,
            string symbol = null, int priceDecimals = -1, int volDecimals = -1)
        {
            int n = epochSeconds.Length;
            if (open.Length != n || high.Length != n || low.Length != n || close.Length != n)
                throw new ArgumentException("All OHLC arrays must match the timestamp count.");
            bool hasVolume = volume != null;
            if (hasVolume && volume.Length != n)
                throw new ArgumentException("volume must match the number of rows.");

            if (priceDecimals < 0)
                priceDecimals = DetectDecimals(Concat(open, high, low, close));
            priceDecimals = Math.Max(0, Math.Min(priceDecimals, 9));

            bool volFloat = false;
            if (hasVolume)
            {
                if (volDecimals < 0) volDecimals = DetectDecimals(volume);
                volDecimals = Math.Max(0, Math.Min(volDecimals, 9));
                volFloat = volDecimals > 0;
            }
            else volDecimals = 0;

            byte[] bin = PackBinary(epochSeconds, open, high, low, close, volume,
                                    priceDecimals, volDecimals, hasVolume, volFloat);
            return BuildUrl(bin, rootUrl, dense, symbol);
        }

        /// <summary>Build a chart URL from in-memory arrays using DateTime timestamps.
        /// DateTimes with Kind=Unspecified are treated as UTC.</summary>
        public static string GenerateUrl(
            DateTime[] times, double[] open, double[] high, double[] low, double[] close,
            double[] volume = null, string rootUrl = null, bool dense = false,
            string symbol = null, int priceDecimals = -1, int volDecimals = -1)
        {
            var epoch = new long[times.Length];
            for (int i = 0; i < times.Length; i++) epoch[i] = ToEpochSeconds(times[i]);
            return GenerateUrl(epoch, open, high, low, close, volume, rootUrl, dense,
                               symbol, priceDecimals, volDecimals);
        }

        /// <summary>Build a chart URL from an OHLCV CSV file (decimals detected from text).</summary>
        public static string GenerateUrlFromFile(
            string path, string rootUrl = null, bool dense = false, string symbol = null)
        {
            Model m = ParseCsvFile(File.ReadAllText(path));
            if (symbol == null) symbol = Path.GetFileNameWithoutExtension(path);
            byte[] bin = PackBinary(m.Ts, m.O, m.H, m.L, m.C, m.Vol,
                                    m.PriceDecimals, m.VolDecimals, m.HasVolume, m.VolFloat);
            return BuildUrl(bin, rootUrl, dense, symbol);
        }

        // ------------------------------------------------------------------ //
        // URL assembly
        // ------------------------------------------------------------------ //
        private static string BuildUrl(byte[] bin, string rootUrl, bool dense, string symbol)
        {
            byte[] comp = DeflateRaw(bin);
            string payload = dense ? DenseEncode(comp) : Base64Url(comp);
            var sb = new StringBuilder();
            sb.Append(rootUrl ?? DefaultRootUrl).Append("#d=").Append(payload);
            if (dense) sb.Append("&e=1");
            if (!string.IsNullOrEmpty(symbol)) sb.Append("&s=").Append(Uri.EscapeDataString(symbol));
            return sb.ToString();
        }

        // ------------------------------------------------------------------ //
        // Stage 1 - columnar binary packing
        // ------------------------------------------------------------------ //
        private static byte[] PackBinary(
            long[] ts, double[] o, double[] h, double[] l, double[] c, double[] vol,
            int priceDecimals, int volDecimals, bool hasVolume, bool volFloat)
        {
            int n = ts.Length;
            double pscale = Math.Pow(10, priceDecimals);
            double vscale = Math.Pow(10, volDecimals);

            bool tsRegular = true;
            long interval = n >= 2 ? ts[1] - ts[0] : 0;
            for (int i = 2; i < n; i++)
                if (ts[i] - ts[i - 1] != interval) { tsRegular = false; break; }

            var outp = new List<byte>();
            outp.Add(0x4F); outp.Add(0x31);            // magic 'O','1'
            outp.Add(1);                                // version
            int flagsIdx = outp.Count;
            outp.Add(0);                                // flags placeholder
            outp.Add((byte)(priceDecimals & 0xFF));
            outp.Add((byte)(volDecimals & 0xFF));
            WriteVarint(outp, n);

            if (tsRegular)
            {
                WriteVarint(outp, n > 0 ? ts[0] : 0);
                WriteVarint(outp, interval > 0 ? interval : 0);
            }
            else
            {
                WriteVarint(outp, ts[0]);
                for (int i = 1; i < n; i++) WriteVarint(outp, Zigzag(ts[i] - ts[i - 1]));
            }

            double[][] cols = { o, h, l, c };
            foreach (var col in cols)
            {
                long prev = 0;
                for (int i = 0; i < n; i++)
                {
                    long v = RoundHalfUp(col[i] * pscale);
                    WriteVarint(outp, Zigzag(v - prev));
                    prev = v;
                }
            }

            bool volDelta = false;
            if (hasVolume)
            {
                var scaled = new long[n];
                for (int i = 0; i < n; i++) scaled[i] = RoundHalfUp((vol[i]) * vscale);
                var raw = new List<byte>();
                var delta = new List<byte>();
                long prev = 0;
                for (int i = 0; i < n; i++)
                {
                    WriteVarint(raw, scaled[i]);
                    WriteVarint(delta, Zigzag(scaled[i] - prev));
                    prev = scaled[i];
                }
                if (delta.Count < raw.Count) { volDelta = true; outp.AddRange(delta); }
                else outp.AddRange(raw);
            }

            int flags = 0;
            if (tsRegular) flags |= 1;
            if (hasVolume) flags |= 2;
            if (volFloat) flags |= 4;
            if (volDelta) flags |= 8;
            outp[flagsIdx] = (byte)flags;

            return outp.ToArray();
        }

        private static void WriteVarint(List<byte> outp, long n)
        {
            while (n >= 128) { outp.Add((byte)((n % 128) + 128)); n /= 128; }
            outp.Add((byte)n);
        }

        private static long Zigzag(long n) { return n >= 0 ? n * 2 : n * -2 - 1; }

        private static long RoundHalfUp(double x) { return (long)Math.Floor(x + 0.5); }

        // ------------------------------------------------------------------ //
        // Stage 2 - raw DEFLATE (no zlib/gzip wrapper)
        // ------------------------------------------------------------------ //
        private static byte[] DeflateRaw(byte[] data)
        {
            using (var ms = new MemoryStream())
            {
                using (var ds = new DeflateStream(ms, CompressionLevel.Optimal, true))
                    ds.Write(data, 0, data.Length);
                return ms.ToArray();
            }
        }

        // ------------------------------------------------------------------ //
        // Stage 3 - URL-safe text encodings
        // ------------------------------------------------------------------ //
        private static string Base64Url(byte[] d)
        {
            return Convert.ToBase64String(d).Replace('+', '-').Replace('/', '_').TrimEnd('=');
        }

        // Whole-buffer base conversion via manual big-number division (matches
        // the BigInt-based encoder in the HTML/Python). A 2-char prefix records
        // the count of leading zero bytes.
        private static string DenseEncode(byte[] data)
        {
            int lead = 0;
            while (lead < data.Length && data[lead] == 0) lead++;
            int len = data.Length - lead;
            var work = new int[len];
            for (int i = 0; i < len; i++) work[i] = data[lead + i];

            var digits = new List<int>();   // least-significant first
            int start = 0;
            while (start < len)
            {
                int rem = 0;
                for (int i = start; i < len; i++)
                {
                    int acc = rem * 256 + work[i];
                    work[i] = acc / DenseN;
                    rem = acc % DenseN;
                }
                digits.Add(rem);
                while (start < len && work[start] == 0) start++;
            }

            var sb = new StringBuilder();
            sb.Append(DenseAlphabet[lead / DenseN]);
            sb.Append(DenseAlphabet[lead % DenseN]);
            for (int i = digits.Count - 1; i >= 0; i--) sb.Append(DenseAlphabet[digits[i]]);
            return sb.ToString();
        }

        // ------------------------------------------------------------------ //
        // Parsing / detection helpers
        // ------------------------------------------------------------------ //
        private static int DetectDecimals(IEnumerable<double> values)
        {
            int d = 0;
            foreach (var v in values)
            {
                int k = 0;
                while (k < 9 && Math.Abs(v * Math.Pow(10, k) - Math.Round(v * Math.Pow(10, k))) > 1e-9) k++;
                if (k > d) d = k;
            }
            return d;
        }

        private static int DecimalsOfToken(string token)
        {
            token = token.Trim();
            int dot = token.LastIndexOf('.');
            if (dot < 0) return 0;
            for (int i = dot + 1; i < token.Length; i++)
                if (!char.IsDigit(token[i])) return 0;
            return token.Length - dot - 1;
        }

        /// <summary>Parse a timestamp string to epoch seconds (naive values = UTC).</summary>
        public static long ToEpochSeconds(string s)
        {
            s = s.Trim();
            if (s.Length > 0 && IsAllDigits(s))
            {
                long num = long.Parse(s, CultureInfo.InvariantCulture);
                return s.Length >= 12 ? (long)Math.Round(num / 1000.0) : num;
            }
            DateTimeOffset dto;
            if (DateTimeOffset.TryParse(s, CultureInfo.InvariantCulture,
                    DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal, out dto))
                return dto.ToUnixTimeSeconds();
            throw new FormatException("Unparseable timestamp: " + s);
        }

        /// <summary>Convert a DateTime to epoch seconds; Kind=Unspecified is treated as UTC.</summary>
        public static long ToEpochSeconds(DateTime dt)
        {
            if (dt.Kind == DateTimeKind.Unspecified) dt = DateTime.SpecifyKind(dt, DateTimeKind.Utc);
            return new DateTimeOffset(dt.ToUniversalTime()).ToUnixTimeSeconds();
        }

        private static bool IsAllDigits(string s)
        {
            foreach (char ch in s) if (ch < '0' || ch > '9') return false;
            return s.Length > 0;
        }

        private static double[] Concat(params double[][] arrays)
        {
            int total = 0;
            foreach (var a in arrays) total += a.Length;
            var res = new double[total];
            int p = 0;
            foreach (var a in arrays) { Array.Copy(a, 0, res, p, a.Length); p += a.Length; }
            return res;
        }

        // ------------------------------------------------------------------ //
        // CSV parsing
        // ------------------------------------------------------------------ //
        private sealed class Model
        {
            public long[] Ts; public double[] O, H, L, C, Vol;
            public int PriceDecimals, VolDecimals;
            public bool HasVolume, VolFloat;
        }

        private static readonly string[] TimeAliases = { "timestamp", "datetime", "dts", "date", "time" };
        private static readonly string[] OpenAliases = { "open", "o" };
        private static readonly string[] HighAliases = { "high", "h" };
        private static readonly string[] LowAliases = { "low", "l" };
        private static readonly string[] CloseAliases = { "close", "c" };
        private static readonly string[] VolAliases = { "volume", "vol", "v" };

        private static Model ParseCsvFile(string text)
        {
            List<string[]> rows = ParseCsv(text);
            if (rows.Count < 2) throw new InvalidDataException("CSV has no data rows.");

            var header = new string[rows[0].Length];
            for (int i = 0; i < header.Length; i++) header[i] = rows[0][i].Trim().ToLowerInvariant();

            int ti = FindCol(header, TimeAliases);
            int oi = FindCol(header, OpenAliases), hi = FindCol(header, HighAliases);
            int li = FindCol(header, LowAliases), ci = FindCol(header, CloseAliases);
            if (ti < 0) throw new InvalidDataException("No time column (timestamp/datetime/dts/date/time).");
            if (oi < 0 || hi < 0 || li < 0 || ci < 0)
                throw new InvalidDataException("Need open, high, low and close columns.");
            int vi = FindCol(header, VolAliases);
            bool hasVolume = vi >= 0;

            int need = Math.Max(Math.Max(Math.Max(ti, oi), Math.Max(hi, li)), Math.Max(ci, hasVolume ? vi : 0));
            var data = new List<string[]>();
            for (int i = 1; i < rows.Count; i++)
                if (rows[i].Length > need && rows[i][ti].Trim().Length > 0) data.Add(rows[i]);
            if (data.Count == 0) throw new InvalidDataException("CSV has no usable data rows.");

            int priceDecimals = 0;
            foreach (var r in data)
                foreach (int idx in new[] { oi, hi, li, ci })
                    priceDecimals = Math.Max(priceDecimals, DecimalsOfToken(r[idx]));
            priceDecimals = Math.Min(priceDecimals, 9);

            int volDecimals = 0; bool volFloat = false;
            if (hasVolume)
            {
                foreach (var r in data)
                {
                    int d = DecimalsOfToken(r[vi]);
                    if (d > 0) volFloat = true;
                    volDecimals = Math.Max(volDecimals, d);
                }
                volDecimals = Math.Min(volDecimals, 9);
            }

            int n = data.Count;
            var m = new Model
            {
                Ts = new long[n],
                O = new double[n], H = new double[n], L = new double[n], C = new double[n],
                Vol = hasVolume ? new double[n] : null,
                PriceDecimals = priceDecimals, VolDecimals = volDecimals,
                HasVolume = hasVolume, VolFloat = volFloat
            };
            for (int i = 0; i < n; i++)
            {
                var r = data[i];
                m.Ts[i] = ToEpochSeconds(r[ti]);
                m.O[i] = ParseDouble(r[oi]); m.H[i] = ParseDouble(r[hi]);
                m.L[i] = ParseDouble(r[li]); m.C[i] = ParseDouble(r[ci]);
                if (hasVolume) m.Vol[i] = TryParseDouble(r[vi]);
            }
            return m;
        }

        private static int FindCol(string[] header, string[] aliases)
        {
            for (int i = 0; i < header.Length; i++)
                foreach (var a in aliases) if (header[i] == a) return i;
            return -1;
        }

        private static double ParseDouble(string s)
        {
            return double.Parse(s.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture);
        }

        private static double TryParseDouble(string s)
        {
            double v;
            return double.TryParse(s.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out v) ? v : 0.0;
        }

        // RFC-4180 parser (mirrors parseCSV in ohlcv_url_chart.html).
        private static List<string[]> ParseCsv(string csv)
        {
            var rows = new List<string[]>();
            var row = new List<string>();
            var field = new StringBuilder();
            bool inQ = false;
            for (int i = 0; i < csv.Length; i++)
            {
                char ch = csv[i];
                if (inQ)
                {
                    if (ch == '"' && i + 1 < csv.Length && csv[i + 1] == '"') { field.Append('"'); i++; }
                    else if (ch == '"') inQ = false;
                    else field.Append(ch);
                }
                else
                {
                    if (ch == '"') inQ = true;
                    else if (ch == ',') { row.Add(field.ToString()); field.Clear(); }
                    else if (ch == '\n')
                    {
                        row.Add(field.ToString()); field.Clear();
                        if (AnyNonEmpty(row)) rows.Add(row.ToArray());
                        row.Clear();
                    }
                    else if (ch != '\r') field.Append(ch);
                }
            }
            if (field.Length > 0 || row.Count > 0)
            {
                row.Add(field.ToString());
                if (AnyNonEmpty(row)) rows.Add(row.ToArray());
            }
            return rows;
        }

        private static bool AnyNonEmpty(List<string> row)
        {
            foreach (var s in row) if (!string.IsNullOrEmpty(s)) return true;
            return false;
        }

#if OHLCV_CLI
        // Console entry point - excluded from NinjaTrader builds (OHLCV_CLI not defined there).
        public static int Main(string[] args)
        {
            if (args.Length == 0)
            {
                Console.Error.WriteLine("Usage: OhlcvUrl <data.csv> [--root URL] [--dense] [--symbol NAME]");
                return 1;
            }
            string csv = args[0], root = DefaultRootUrl, symbol = null;
            bool dense = false;
            for (int i = 1; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--root": root = args[++i]; break;
                    case "--dense": dense = true; break;
                    case "--symbol": symbol = args[++i]; break;
                    default: Console.Error.WriteLine("Unknown arg: " + args[i]); return 1;
                }
            }
            try { Console.WriteLine(GenerateUrlFromFile(csv, root, dense, symbol)); return 0; }
            catch (Exception ex) { Console.Error.WriteLine("Error: " + ex.Message); return 2; }
        }
#endif
    }
}
